rpcserver.go

```diff
@@ -2544,226 +2574,478 @@ func marshalOutputType(outputType tappsbt.VOutputType) (taprpc.OutputType,
+			return e.Status == address.StatusCompleted, nil
+
+		default:
+			return false, fmt.Errorf("unknown event type: %T", e)
+		}
+	}
+
+	return handleEvents[bool, *tapdevrpc.ReceiveAssetEvent](
+		r.cfg.AssetCustodian, ntfnStream, marshaler, filter, r.quit,
+	)
+}
+
+// SubscribeReceiveEvents registers a subscription to the event notification
+// stream which relates to the asset receiving process.
+func (r *rpcServer) SubscribeReceiveEvents(
+	req *taprpc.SubscribeReceiveEventsRequest,
+	ntfnStream receiveEventStream) error {
+
+	tapParams := address.ParamsForChain(r.cfg.ChainParams.Name)
+
+	// We just decode the address to make sure it's valid. But any
+	// comparison for filtering happens on the string representation, as
+	// that's easily comparable
+	var addrString string
+	if req.FilterAddr != "" {
+		addr, err := address.DecodeAddress(req.FilterAddr, &tapParams)
+		if err != nil {
+			return fmt.Errorf("error decoding address: %w", err)
+		}
+
+		addrString, err = addr.EncodeAddress()
+		if err != nil {
+			return fmt.Errorf("error encoding address: %w", err)
+		}
+	}
+
+	marshaler := func(event fn.Event) (*taprpc.ReceiveEvent, error) {
+		e, ok := event.(*tapgarden.AssetReceiveEvent)
+		if !ok {
+			return nil, fmt.Errorf("invalid event type: %T", event)
+		}
 
-			err = ntfnStream.Send(rpcEvent)
+		rpcAddr, err := marshalAddr(&e.Address, r.cfg.TapAddrBook)
+		if err != nil {
+			return nil, fmt.Errorf("error marshaling addr: %w", err)
+		}
+
+		rpcStatus, err := marshalAddrEventStatus(e.Status)
+		if err != nil {
+			return nil, fmt.Errorf("error marshaling status: %w",
+				err)
+		}
+
+		return &taprpc.ReceiveEvent{
+			Timestamp: e.Timestamp().UnixMicro(),
+			Address:   rpcAddr,
+			Outpoint:  e.OutPoint.String(),
+			Status:    rpcStatus,
+		}, nil
+	}
+
+	filter := func(event fn.Event) (bool, error) {
+		var (
+			eventAddrString string
+			err             error
+		)
+
+		switch e := event.(type) {
+		case *tapgarden.AssetReceiveEvent:
+			eventAddrString, err = e.Address.EncodeAddress()
 			if err != nil {
-				return fmt.Errorf("failed to RPC stream send "+
-					"event: %w", err)
+				return false, fmt.Errorf("error encoding "+
+					"address: %w", err)
 			}
 
-		// Handle the case where the RPC stream is closed by the
-		// client.
-		case <-ntfnStream.Context().Done():
-			// Don't return an error if a normal context
-			// cancellation has occurred.
-			isCanceledContext := errors.Is(
-				ntfnStream.Context().Err(), context.Canceled,
-			)
-			if isCanceledContext {
-				return nil
-			}
+		case *proof.BackoffWaitEvent:
+			// We're not interested in any backoff events.
+			return false, nil
 
-			return ntfnStream.Context().Err()
+		default:
+			return false, fmt.Errorf("unknown event type: %T", e)
+		}
 
-		// Handle the case where the RPC server is shutting down.
-		case <-r.quit:
-			return nil
+		// If we're not filtering on a specific address, we return all
+		// events.
+		if addrString == "" {
```

Comment:
super-nit: could be a one-line with `(addrString == "" || eventAddrString == addrString), nil`

Reply:
True. But I try to avoid chaining too many boolean operators together, as they can sometimes be hard to read or understand at a quick glance. I'd rather e a bit more explicit here.